File: web/src/entities/navigation.ts
export type GlobalSearchHit = {
  target_type: "document" | "image" | "memory";
  target_id: string;
  title: string;
  excerpt: string;
  score: number;
  tags: string[];
  metadata: Record<string, string | number | null>;
};

export type SearchResponse = {
  query: string;
  documents: GlobalSearchHit[];
  images: GlobalSearchHit[];
  memories: GlobalSearchHit[];
};

export type Favorite = {
  id: string;
  target_type: string;
  target_id: string;
  snapshot: Record<string, unknown> | null;
  created_at: string;
};

export type ManagedTag = {
  id: string;
  name: string;
  color: string;
  source: string;
  document_count: number;
  image_count: number;
};

export type DailyReview = {
  id: string;
  review_date: string;
  content: string;
  stats: Record<string, number>;
  created_at: string;
  updated_at: string;
};

export type DashboardData = {
  counts: Record<string, number>;
  tag_distribution: Array<{ name: string; value: number }>;
  memory_trend: Array<{ date: string; value: number }>;
  community_distribution: Array<{ name: string; value: number }>;
  observability: ObservabilitySummary;
};

export type ObservabilitySummary = {
  period_days: number;
  traces: number;
  failed_traces: number;
  success_rate: number;
  avg_duration_ms: number;
  tool_calls: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  token_trend: Array<{ date: string; input_tokens: number; output_tokens: number }>;
  tool_distribution: Array<{ name: string; value: number }>;
};

export type AgentTrace = {
  id: string;
  conversation_id: string;
  user_message_id: string;
  assistant_message_id: string | null;
  status: "running" | "completed" | "failed";
  mode: string | null;
  question: string;
  started_at: string;
  finished_at: string | null;
  duration_ms: number | null;
  tool_call_count: number;
  citation_count: number;
  error_message: string | null;
};

export type AgentSpan = {
  id: string;
  parent_span_id: string | null;
  kind: string;
  name: string;
  status: string;
  started_at: string;
  finished_at: string;
  duration_ms: number;
  input_summary: Record<string, unknown> | null;
  output_summary: Record<string, unknown> | null;
  error_message: string | null;
};

export type ModelUsage = {
  id: string;
  operation: string;
  provider: string;
  model: string;
  status: string;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  duration_ms: number | null;
  error_message: string | null;
};

export type RetrievalSnapshot = {
  id: string;
  tool_call_id: string;
  tool_name: string;
  query: string;
  hit_count: number;
  duration_ms: number;
  status: string;
  citations: Array<Record<string, unknown>> | null;
  result_metadata: Record<string, unknown> | null;
  top_score: number | null;
};

export type AgentTraceDetail = AgentTrace & {
  spans: AgentSpan[];
  model_usages: ModelUsage[];
  retrieval_snapshots: RetrievalSnapshot[];
};

File: api/app/api/schemas/navigation.py
"""Global search, favorites, tags, daily review, and dashboard schemas."""

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.api.schemas.observability import ObservabilitySummary


class GlobalSearchRequest(BaseModel):
    query: str = Field(min_length=1, max_length=2000)
    top_k: int = Field(default=8, ge=1, le=20)
    min_score: float = Field(default=0.25, ge=0.0, le=1.0)


class GlobalSearchHit(BaseModel):
    target_type: Literal["document", "image", "memory"]
    target_id: str
    title: str
    excerpt: str
    score: float
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, object] = Field(default_factory=dict)


class GlobalSearchResponse(BaseModel):
    query: str
    documents: list[GlobalSearchHit]
    images: list[GlobalSearchHit]
    memories: list[GlobalSearchHit]


class FavoriteCreate(BaseModel):
    target_type: Literal["document", "image", "memory", "message"]
    target_id: str = Field(min_length=1, max_length=64)
    snapshot: dict[str, object] | None = None


class FavoriteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    target_type: str
    target_id: str
    snapshot: dict[str, object] | None
    created_at: datetime


class TagCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    color: str | None = Field(default=None, pattern=r"^#[0-9A-Fa-f]{6}$")


class TagUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=64)
    color: str | None = Field(default=None, pattern=r"^#[0-9A-Fa-f]{6}$")


class TagManagementResponse(BaseModel):
    id: uuid.UUID
    name: str
    color: str
    source: str
    document_count: int
    image_count: int


class DailyReviewResponse(BaseModel):
    id: uuid.UUID
    review_date: date
    content: str
    stats: dict[str, object]
    created_at: datetime
    updated_at: datetime


class DashboardResponse(BaseModel):
    counts: dict[str, int]
    tag_distribution: list[dict[str, object]]
    memory_trend: list[dict[str, object]]
    community_distribution: list[dict[str, object]]
    observability: ObservabilitySummary
